### ISSUE TITLE
twisted.names.client doesn't close '/etc/resolv.conf'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6216]
|<img alt="duffy151's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| duffy151 reported|
|-|-|
|Trac ID|trac#6216|
|Type|defect|
|Created|2012-12-20 22:14:22Z|
|Branch|https://github.com/twisted/twisted/tree/close-resolve-conf-6216|

in maybeParseConfig(), /etc/resolv.conf is opened but never closed. In PyPy this can lead to out of file descriptor errors until the gc collects them.

Attachments:

* [resolvconf-close-6216.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46d/46d5f8969055c735a53a5a1f4f558ba1ab39caff/fa8f003b556dfec4928a1e47f7e5895125380cd3.patch) (493 bytes) - added by ephess on 2012-12-21 22:05:05Z - 
* [resolvconf-close-6216.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46d/46d5f8969055c735a53a5a1f4f558ba1ab39caff/9f01d4e77d184d0a573595376f4e050aff1ce2c2.patch) (2777 bytes) - added by ephess on 2013-01-03 20:11:27Z - Revised patch addressing concerns raised in first review

<details><summary>Searchable metadata</summary>

```
trac-id__6216 6216
type__defect defect
reporter__duffy151 duffy151
priority__normal normal
milestone__ 
branch__branches_close_resolve_conf_6216 branches/close-resolve-conf-6216
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1356041662000000 1356041662000000
changetime__1357467785000000 1357467785000000
version__None None
owner__glyph glyph
cc__ephess cc__borko
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ephess commented|
|-|-|

Patch with fix uploaded.

Reviewer: Code is already exercised by a unit test and this isn't something I'd write a specific test for normally. Let me know if this approach is OK?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to ephess|
|-|-|

Thanks for working on this.  Generally, all functionality should be unit tested.  You're fixing a bug here, which causes real problems on PyPy: anyone who cares about not having these problems will care about there being a test that demonstrates the problems are not re-introduced.

I can see two possible testing approaches for this.

First, hook into the file being opened somehow, so that you can assert that it is also closed at some point.  For example, you could add a (private) helper method to `client.Resolver` for opening files, change the code to use it, and then override it in the unit tests to hand back a `StringIO` instance which you later check to make sure close has been called on it (somewhat like what is done in `twisted.web.test.test_static.StaticProducerTests.test_stopProducingClosesFile`).

Second, take advantage of the `ResourceWarning` which some versions of Python emit when a file is closed by the garbage collector.  This involves working on the BuildBot configuration to recognize these warnings as errors and fail a build whenever a new instance is introduced.  Then you can fix this problem (which should be causing a `ResourceWarning` to be emitted) and if anyone ever re-introduces it, BuildBot will report a failed build.  This is probably the harder of the two options, and not quite as nice as the first since it won't report problems to developers locally, it will only report them after they trigger a build on BuildBot.

Also, separately, you should add a [ReviewProcess#Newsfiles news fragment] for this fix.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ephess commented|
|-|-|

Thanks for the feedback exarkun - have gone with the first option. Do you think it would be worth while opening up a second ticket to make the change in Buildbot, as this would also pick up any other parts of the code base that are doing this also?

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ephess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

(In [36732]) Branching to 'close-resolve-conf-6216'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @glyph|
|-|-|

Hi ephess,

There are a couple of minor issues with this patch, but I'll address them quickly and then merge.

  1. The standard we use for docstrings in tests is "Foo does bar." and not "Ensure that we verify that foo indeed correctly does bar.".  Hopefully that slightly exaggerated example makes it clear why ;-).
  2. You should generally avoid `patch` in tests, since it manipulates global state that either the test framework itself or other tests could be using.  If you're going to use `patch` to test this, you don't actually need to add the `_openFile` stuff, since you can just `patch(twisted.names.client, "FilePath", MyFakeFilePath)`.  If you are going to add the `_openFile` hook, then you can easily just pass it in rather than having to patch it up.  So, in a sense, this test is the worst of both worlds.

Thanks for your contribution!

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

After making those changes and forcing a build, I noticed that my change introduced a [test failure on python 3](http://buildbot.twistedmatrix.com/builders/python-3.3-tests/builds/344/steps/shell_1/logs/stdio); I guess we need to use `FilePath` after all!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set status to `closed`|
|-|-|

(In [36746]) Merge close-resolve-conf-6216

Author: ephess

Reviewer: glyph

Fixes: [#6216](https://github.com/twisted/twisted/issues/6216)

Do not depend on the garbage collector to close /etc/resolv.conf when twisted.names.client is done reading it.
